File: backend/src/main/java/fr/pedalons/dto/routes/response/RouteDetailDto.java
package fr.pedalons.dto.routes.response;

import fr.pedalons.common.TsidUtils;
import fr.pedalons.domain.route.Route;
import fr.pedalons.dto.comments.response.CommentCounts;
import fr.pedalons.dto.common.GeoJsonPoint;
import fr.pedalons.dto.common.asset.MediaDto;
import fr.pedalons.dto.publications.response.TeamPublicationDto;
import fr.pedalons.dto.routes.request.GeometryOptions;
import fr.pedalons.dto.users.response.PublicUserDto;
import fr.pedalons.dto.validation.ValidateSchema;
import fr.pedalons.enums.SurfaceType;
import fr.pedalons.enums.Visibility;
import fr.pedalons.service.asset.AssetService;
import java.time.Instant;
import java.util.List;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.geolatte.geom.G2D;
import org.geolatte.geom.Point;
import org.jspecify.annotations.Nullable;

/**
 * Route DTO for detail view with full information.
 */
@Schema(description = "Detailed route information")
@ValidateSchema
public record RouteDetailDto(
    @Schema(description = "Route ID (TSID)", required = true) String id,
    @Schema(description = "Route slug", required = true) String slug,
    @Schema(description = "Team", required = true) TeamPublicationDto team,
    @Schema(description = "Route name", required = true) String name,
    @Schema(description = "Media", required = true) MediaDto media,
    @Schema(description = "Distance in meters", required = true) Float distance,
    @Schema(description = "Total elevation gain in meters", required = true) Float elevationGain,
    @Schema(description = "Total elevation loss in meters", required = true) Float elevationLoss,
    @Schema(description = "Surface type", required = true) SurfaceType surfaceType,
    @Schema(description = "Whether the route is public", required = true) Visibility visibility,
    @Schema(implementation = GeoJsonPoint.class) Point<G2D> start,
    @Schema(implementation = GeoJsonPoint.class) Point<G2D> end,
    @Schema(description = "Creator user", required = true) PublicUserDto createdBy,
    @Schema(description = "Creation timestamp", required = true) Instant createdAt,
    @Schema(description = "Last update timestamp", required = true) Instant updatedAt,
    @Schema(description = "Tracks", required = true) List<TrackDto> tracks,
    @Schema(description = "Waypoints", required = true) List<WaypointDto> waypoints,
    @Schema(description = "Whether the route is soft-deleted", required = true) boolean deleted,
    @Nullable
        @Schema(
            description =
                "Number of comments, replies included. Absent when the caller may not read the"
                    + " comments of this route — comments are members-only, so an outsider is told"
                    + " nothing, not even zero.")
        Integer commentCount) {

  public static RouteDetailDto from(Route route, AssetService assetService) {
    return from(route, assetService, GeometryOptions.FULL, CommentCounts.NONE);
  }

  public static RouteDetailDto from(
      Route route, AssetService assetService, GeometryOptions geometry) {
    return from(route, assetService, geometry, CommentCounts.NONE);
  }

  /** Same detail, with every track decimated as {@code geometry} asks. */
  public static RouteDetailDto from(
      Route route,
      AssetService assetService,
      GeometryOptions geometry,
      CommentCounts commentCounts) {
    return new RouteDetailDto(
        TsidUtils.toString(route.getId()),
        route.getSlug(),
        TeamPublicationDto.from(route.getTeam()),
        route.getName(),
        MediaDto.from(route, assetService),
        route.getDistance(),
        route.getElevationGain(),
        route.getElevationLoss(),
        route.getSurfaceType(),
        route.getVisibility(),
        route.getStart(),
        route.getEnd(),
        PublicUserDto.from(route.getCreatedBy()),
        route.getCreatedAt(),
        route.getUpdatedAt(),
        route.getTracks().stream().map(track -> TrackDto.from(track, geometry)).toList(),
        route.getWaypoints().stream().map(WaypointDto::from).toList(),
        route.isDeleted(),
        commentCounts.forEntity(route.getId()));
  }
}

File: frontend/src/api/dto/routeDetailDto.ts
import type { ElevationProfileDto } from './elevationProfileDto.ts'
import type { GeoJsonPoint } from './geoJsonPoint.ts'
import type { Instant } from './instant.ts'
import type { MediaDto } from './mediaDto.ts'
import type { PublicUserDto } from './publicUserDto.ts'
import type { SurfaceType } from './surfaceType.ts'
import type { TeamPublicationDto } from './teamPublicationDto.ts'
import type { TrackDto } from './trackDto.ts'
import type { Visibility } from './visibility.ts'
import type { WaypointDto } from './waypointDto.ts'

/**
 * Detailed route information
 */
export interface RouteDetailDto {
  /** Route ID (TSID) */
  id: string
  /** Route slug */
  slug: string
  /** Team */
  team: TeamPublicationDto
  /** Route name */
  name: string
  /** Media */
  media: MediaDto
  /** Distance in meters */
  distance: number
  /** Total elevation gain in meters */
  elevationGain: number
  /** Total elevation loss in meters */
  elevationLoss: number
  /** Surface type */
  surfaceType: SurfaceType
  /** Whether the route is public */
  visibility: Visibility
  start?: GeoJsonPoint
  end?: GeoJsonPoint
  /** Creator user */
  createdBy: PublicUserDto
  /** Creation timestamp */
  createdAt: Instant
  /** Last update timestamp */
  updatedAt: Instant
  /** Tracks */
  tracks: TrackDto[]
  /** Waypoints */
  waypoints: WaypointDto[]
  /** Whether the route is soft-deleted */
  deleted: boolean
  /** Number of comments, replies included. Absent when the caller may not read the comments of this route — comments are members-only, so an outsider is told nothing, not even zero. */
  commentCount?: number
  /** Sampled elevation profile of the route. Absent unless explicitly requested (the bulk route endpoint's 'elevation' flag) — computing and serialising it costs nothing to skip, so every other caller of this DTO gets exactly what it got before this field existed. */
  elevationProfile?: ElevationProfileDto
}
